linalg/tensor.hpp

```diff
@@ -0,0 +1,1771 @@
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] scale The scaling factor
+ * @param[in] A The tensor to be scaled
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+MFEM_HOST_DEVICE auto operator*(S scale, const tensor<T, n...>& A) ->
+tensor<decltype(S {} * T{}), n...>
+{
+   tensor<decltype(S{} * T{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = scale * A[i];
+   }
+   return C;
+}
+
+/**
+ * @brief multiply a tensor by a scalar value
+ * @tparam S the scalar value type. Must be arithmetic (e.g. float, double, int) or a dual number
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The tensor to be scaled
+ * @param[in] scale The scaling factor
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+MFEM_HOST_DEVICE auto operator*(const tensor<T, n...>& A, S scale) ->
+tensor<decltype(T {} * S{}), n...>
+{
+   tensor<decltype(T{} * S{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = A[i] * scale;
+   }
+   return C;
+}
+
+/**
+ * @brief divide a scalar by each element in a tensor
+ * @tparam S the scalar value type. Must be arithmetic (e.g. float, double, int) or a dual number
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] scale The numerator
+ * @param[in] A The tensor of denominators
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+auto operator/(S scale, const tensor<T, n...>& A) ->
+tensor<decltype(S {} * T{}), n...>
+{
+   tensor<decltype(S{} * T{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = scale / A[i];
+   }
+   return C;
+}
+
+/**
+ * @brief divide a tensor by a scalar
+ * @tparam S the scalar value type. Must be arithmetic (e.g. float, double, int) or a dual number
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The tensor of numerators
+ * @param[in] scale The denominator
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+auto operator/(const tensor<T, n...>& A, S scale) ->
```

Comment:
```suggestion
MFEM_HOST_DEVICE auto operator/(const tensor<T, n...>& A, S scale) ->
```

Reply:
there are a handful of other incidences where these functions are missing their annotations, I believe almost all of the functions in here should be `__host__ __device__` (maybe with a few exceptions like the `std::ostream` operator<< overload).